### ISSUE TITLE
[bug] Logit scaling incorrect for triton and torch cross-entropy

### ISSUE THREAD

**Issue by jlamypoirier** [ServiceNow/Fast-LLM#228]
# 🐞 Describe the Bug

Triton implementation of logit scaling for cross-entropy is incorrect:
* Re-loaded logits are not scaled https://github.com/ServiceNow/Fast-LLM/blame/main/fast_llm/functional/triton/cross_entropy.py#L36
* Scaling gradient is applied at the wrong place (on exp logits instead of final gradient) https://github.com/ServiceNow/Fast-LLM/blame/main/fast_llm/functional/triton/cross_entropy.py#L48

Also the torch implementation crashes with logit scaling or with `grad_output=None`. https://github.com/ServiceNow/Fast-LLM/blob/main/fast_llm/functional/cross_entropy.py#L11.

The bugs have been around since loss scaling was first implemented in e933bd6, so could have impacted experiments.

Found those while adapting cross-entropy for knowledge distillation #214. I already made fixes and tests, I'll include in the PR unless there is an immediate need.

**Comment by tscholak** [comment]
thanks for catching this!
we never used loss scaling, luckily.
